File: Programmers/체육복.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    int answer = 0;
    int *arr = new int[n] {0, };
    
    for (int value : lost) arr[value - 1]--;
    for (int value : reserve) arr[value - 1]++;
    for (int i=0; i<n; i++) {
        if (arr[i] >= 1) {
            if (i != 0 && arr[i-1] == -1) {
                arr[i]--;
                arr[i-1]++;
            } else if (i != n-1 && arr[i+1] == -1) {
                arr[i]--;
                arr[i+1]++;
            }
        }
    }
        
    for (int i=0; i<n; i++) {
        if (arr[i] >= 0) answer++;
    }
    
    return answer;
}

File: Jay/programmers/src/체육복.kt
fun main() {
    val exam = hashSetOf<Int>(4, 1, 3)
    exam.forEach {
        println(it)
    }
}

class Solution체육복 {
    fun solution(n: Int, lost: IntArray, reserve: IntArray): Int {
        val countList = MutableList<Int>(n) { 1 }

        lost.sort()
        reserve.sort()

        reserve.forEach {
            countList[it - 1] += 1
        }
        lost.forEach {
            countList[it - 1] -= 1
        }

        countList.forEach {
            print(it)
        }

        run {
            lost.forEach loop@{ lostIdx ->
                reserve.forEach { reserveIdx ->
                    if (countList[lostIdx - 1] == 1) {
                        return@loop
                    }
                    if (lostIdx - 1 == reserveIdx || lostIdx + 1 == reserveIdx) {
                        if (countList[lostIdx - 1] == 0 && countList[reserveIdx - 1] == 2) {
                            countList[lostIdx - 1] = 1
                            countList[reserveIdx - 1] = 1
                            return@run
                        }
                    }
                }
            }
        }
        return countList.count { it > 0 }
    }
}

class Solution체육복2 {
    fun solution(n: Int, lost: IntArray, reserve: IntArray): Int {

        var answer = n
        var lostSet = lost.toSet() - reserve.toSet()
        var reserveSet = (reserve.toSet() - lost.toSet()) as MutableSet

        for (i in lostSet) {
            when {
                i + 1 in reserveSet -> reserveSet.remove(i + 1)
                i - 1 in reserveSet -> reserveSet.remove(i - 1)
                else -> answer--
            }
        }
        return answer
    }
}